### PR TITLE
Added endpoint for fetching single rule

### PR DESCRIPTION
**Pull request by bond95**
# Description

Changed DB, added endpoint.

Fixes https://github.com/RedHatInsights/insights-results-aggregator/issues/896

## Type of change

Please delete options that are not relevant.

- New feature (non-breaking change which adds functionality)
- Documentation update

## Testing steps

Run tests

### CODE REVIEW

File: openapi.json [hunk 1]
```diff
@@ -263,6 +263,103 @@
         ]
       }
     },
+    "/organizations/{orgId}/clusters/{clusterId}/users/{userId}/report/{ruleID}": {
```
Comment (quarckster) [inline openapi.json:266]: Don't we already have it in https://github.com/RedHatInsights/insights-results-smart-proxy/blob/master/openapi.json#L355

**Comment by tisnik** [comment]
Please rebase @bond95

File: openapi.json (hunk 1 above)
Comment (bond95) [reply openapi.json:266]: Yes, but in smart proxy. Now we need to have endpoint for single rule in aggregator.

Comment (quarckster) [reply openapi.json:266]: @bond95 I successfully fetched a report for a single rule from CI and QA. Why is this PR needed now if everything works already?

File: consumer/benchmark_test.go [hunk 2]
```diff
@@ -104,17 +104,12 @@ func getMessagesFromDir(b *testing.B, dataDir string) []string {
 				helpers.FailOnError(b, err)
 
 				zerolog.SetGlobalLevel(zerolog.Disabled)
-				parsedMessage, err := consumer.ParseMessage(fileBytes)
```
Comment (Sergey1011010) [inline consumer/benchmark_test.go:107]: why is it deleted?

File: metrics/metrics_test.go [hunk 3]
```diff
@@ -41,6 +42,25 @@ const (
 )
 
 var (
+	report3Rules = types.KafkaReport{
```
Comment (Sergey1011010) [inline metrics/metrics_test.go:45]: why not in testdata?

File: migration/mig_0012_add_rule_id_field.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"database/sql"
+
+	"github.com/RedHatInsights/insights-results-aggregator/types"
+)
+
+var mig0012AddRuleIDField = Migration{
+	StepUp: func(tx *sql.Tx, _ types.DBDriver) error {
+		_, err := tx.Exec(`
+			ALTER TABLE report ADD COLUMN rule_id VARCHAR NOT NULL
```
Comment (Sergey1011010) [inline migration/mig_0012_add_rule_id_field.go:30]: what about shitty dbs like sqlite?

Comment (Sergey1011010) [reply migration/mig_0012_add_rule_id_field.go:30]: and what is this column for?

File: server/server_test.go [hunk 5]
```diff
@@ -47,6 +48,26 @@ var config = server.Configuration{
 	Auth:        false,
 }
 
+var report3Rules = types.KafkaReport{
```
Comment (Sergey1011010) [inline server/server_test.go:51]: why not in testdata?

File: metrics/metrics_test.go (hunk 3 above)
Comment (bond95) [reply metrics/metrics_test.go:45]: Because WIP

File: migration/mig_0013_add_rule_table.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+		if driver == types.DBDriverPostgres {
+			_, err = tx.Exec(`
+				ALTER TABLE report DROP COLUMN report
+			`)
+		} else {
+			err = downgradeTable(
+				tx,
+				clusterReportTable,
+				`
+				CREATE TABLE report (
```
Comment (Sergey1011010) [inline migration/mig_0013_add_rule_table.go:39]: It's gonna fuck up our database, you need to either clear the table completely(since we're caching the latest results, reports will fill the DB pretty quickly), or migrate rules to a new table

Comment (Sergey1011010) [reply migration/mig_0013_add_rule_table.go:39]: Second is actually hard and I don't see a reason to do that

File: openapi.json (hunk 1 above)
Comment (Sergey1011010) [reply openapi.json:266]: It's better to follow the convention and use `/rules/{ruleId}`

File: openapi.json [hunk 7]
```diff
@@ -263,6 +263,103 @@
...
+      "get": {
+        "summary": "Returns the latest rule report for the given organization and cluster which contains information about rules that were hit by the cluster.",
```
Comment (Sergey1011010) [inline openapi.json:268]: Summary is outdated, isn't it?

File: openapi.json [hunk 8]
```diff
@@ -263,6 +263,103 @@
...
+            "in": "path",
+            "required": true,
+            "schema": {
+              "type": "string"
+            }
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Latest available rule report for the given organization, cluster and rule id combination. Returns rule and it description that were hit by the cluster.",
```
Comment (Sergey1011010) [inline openapi.json:314]: as well as the description here

File: storage/storage.go [hunk 9]
```diff
@@ -230,6 +232,13 @@ func (storage DBStorage) Init() error {
 		storage.clustersLastChecked[clusterName] = lastChecked
 	}
 
+	if storage.dbDriverType == types.DBDriverSQLite3 {
+		_, err = storage.connection.Exec("PRAGMA foreign_keys = OFF;")
```
Comment (Sergey1011010) [inline storage/storage.go:236]: We don't want to disable foreign_keys, they were enabled in helpers mock DB on purpose

File: storage/storage.go [hunk 10]
```diff
@@ -386,34 +461,48 @@ func commaSeparatedStrToTags(str string) []string {
 func (storage DBStorage) getReportUpsertQuery() (string, error) {
```
Comment (Sergey1011010) [inline storage/storage.go:461]: why does this function returns queries operating on different tables for different DB engines?

File: storage/storage.go [hunk 11]
```diff
@@ -424,14 +513,59 @@ func (storage DBStorage) WriteReportForCluster(
 		return err
 	}
 
+	deleteQuery := "DELETE FROM rule WHERE org_id = $1 AND cluster_id = $2;"
```
Comment (Sergey1011010) [inline storage/storage.go:516]: Why do you delete if you have upsert in next few lines?

File: storage/storage.go [hunk 12]
```diff
@@ -424,14 +513,59 @@ func (storage DBStorage) WriteReportForCluster(
...
+		log.Err(err).Msgf("Unable to remove previous cluster reports (org: %v, cluster: %v)", orgID, clusterName)
+		_ = tx.Rollback()
+		return err
+	}
+
+	// Perform the report upsert.
+	reportedAtTime := time.Now()
+
+	for _, rule := range rules {
+		ruleID := fmt.Sprintf("%v|%v", rule.Module, rule.ErrorKey)
```
Comment (Sergey1011010) [inline storage/storage.go:528]: Why do you use ruleID combined with ruleErrorKey by '|' sign? I think it's gonna confuse us a lot and postgres has no problems with composite primary keys.

Comment (Sergey1011010) [reply storage/storage.go:528]: It can't cause problems since it's python module's name and  '|' can't be used there, but it looks weird. But consider it as a nitpick :)

**Comment by Sergey1011010** [comment]
Finally, we have a was to not loop trough all the rules, thanks

**Comment by Sergey1011010** [comment]
> ATM it is the slowest item in the whole pipeline and can't be scaled up.

Well, it can be, we even use kafka's groups now, but it requires few days to implement :)

File: openapi.json (hunk 7 above)
Comment (bond95) [reply openapi.json:268]: No, it's fine

File: openapi.json (hunk 8 above)
Comment (bond95) [reply openapi.json:314]: Here as well

File: storage/storage.go (hunk 9 above)
Comment (bond95) [reply storage/storage.go:236]: It's done because removing column in SQLite in migrations will not work, otherwise we need to recreate whole DB from scratch. Solutions that I see is or disable foreign keys for SQLite or remove SQLite support completely.

File: storage/storage.go (hunk 10 above)
Comment (bond95) [reply storage/storage.go:461]: IDK I didn't add this function, but I assume it's because code is clear with this function

File: storage/storage.go (hunk 11 above)
Comment (bond95) [reply storage/storage.go:516]: Because we should remove previous results before adding new one

File: storage/storage.go (hunk 12 above)
Comment (bond95) [reply storage/storage.go:528]: Because there can be different reports with same `Module`, but different `ErrorKey`, thus only combination of this two values might be perceive as unique key

Comment (Sergey1011010) [reply storage/storage.go:528]: Yes, and this is what composite primary keys are for

File: storage/storage.go (hunk 11 above)
Comment (Sergey1011010) [reply storage/storage.go:516]: What's the purpose of upsert then?

File: storage/storage.go (hunk 9 above)
Comment (Sergey1011010) [reply storage/storage.go:236]: Then you can turn them off temporarily in the migration.

> remove SQLite support completely.

I like this idea, but it's a decision we need to discuss as a team

File: openapi.json (hunk 7 above)
Comment (Sergey1011010) [reply openapi.json:268]: > which contains information about rules that were hit by the cluster

I thought the endpoint returns one rule

**Comment by joselsegura** [comment]
The overall code seems ok for me, except for some nitpicks. My main concerns about this change are the following:

- New Kafka records parsing procedure will add a lot of more semantic to the stored information, and that is very good, but as we are not storing the whole record message anywhere
- I didn't see any code in the migration code to move current data in the database (in raw format) to the new structure.

File: storage/storage.go (hunk 11 above)
Comment (bond95) [reply storage/storage.go:516]: Insert data to DB

File: openapi.json (hunk 7 above)
Comment (bond95) [reply openapi.json:268]: Yes, `latest rule report`

File: server/server.go [hunk 13]
```diff
@@ -173,10 +168,68 @@ func (server *HTTPServer) readReportForCluster(writer http.ResponseWriter, reque
...
+
+	var ruleDetails json.RawMessage
+	err = json.Unmarshal([]byte(report), &ruleDetails)
+	if err != nil {
+		log.Error().Err(err).Msg("Unable to parse cluster rule report")
+		handleServerError(writer, err)
+		return
+	}
+
+	ruleIDSplitted := strings.Split(string(ruleID), "|")
```
Comment (Sergey1011010) [inline server/server.go:218]: Does it even work? You already splitted rule id in `readRuleIDWithErrorKey`

File: server/server.go [hunk 14]
```diff
@@ -173,10 +168,68 @@ func (server *HTTPServer) readReportForCluster(writer http.ResponseWriter, reque
...
+	reportRule := types.RuleOnReport{
+		TemplateData: string(ruleDetails),
+		Module:       types.RuleID(ruleIDSplitted[0]),
+		ErrorKey:     types.ErrorKey(ruleIDSplitted[1]),
```
Comment (Sergey1011010) [inline server/server.go:222]: Isn't it gonna panic?

File: storage/storage.go [hunk 15]
```diff
@@ -325,42 +335,108 @@ func (storage DBStorage) GetOrgIDByClusterID(cluster types.ClusterName) (types.O
...
+	report := make([]types.RuleOnReport, 0)
+	for rows.Next() {
+		var details string
+		var ruleID string
+		err := rows.Scan(&details, &ruleID)
+		if err != nil {
+			log.Error().Err(err).Msg("ReportListForCluster")
+			return report, err
+		}
+		var ruleDetails json.RawMessage
```
Comment (Sergey1011010) [inline storage/storage.go:348]: Why do you need to unmarshal to json.RawMessage and then convert to string when the DB field is already a string?

File: storage/storage_test.go [hunk 16]
```diff
@@ -276,8 +291,23 @@ func TestDBStorageWriteReportForClusterExecError(t *testing.T) {
 	_, err := connection.Exec(query)
 	helpers.FailOnError(t, err)
 
+	query = `
+		CREATE TABLE rule_hit (
+			org_id          INTEGER NOT NULL,
+			cluster_id      VARCHAR NOT NULL,
+			rule_id         VARCHAR NOT NULL,
+			error_key       VARCHAR NOT NULL,
+			report          VARCHAR NOT NULL,
```
Comment (Sergey1011010) [inline storage/storage_test.go:300]: If I'm reading the code correctly, you put template data here. Why is it called report then?

File: storage/storage.go [hunk 17]
```diff
@@ -440,22 +572,17 @@ func (storage DBStorage) WriteReportForCluster(
...
+	err = storage.updateReport(tx, orgID, clusterName, report, rules, lastCheckedTime, kafkaOffset)
 	if err != nil {
-		log.Err(err).Msgf("Unable to upsert the cluster report (org: %v, cluster: %v)", orgID, clusterName)
-		_ = tx.Rollback()
 		return err
 	}
 
 	storage.clustersLastChecked[clusterName] = lastCheckedTime
 	metrics.WrittenReports.Inc()
-	return tx.Commit()
```
Comment (Sergey1011010) [inline storage/storage.go:458]: Why did you remove commit here?

File: server/server.go [hunk 18]
```diff
@@ -206,6 +206,14 @@ func (server *HTTPServer) readSingleRule(writer http.ResponseWriter, request *ht
 		return
 	}
 
+	var ruleDetails json.RawMessage
```
[the remark this answers is not in the document]
Comment (bond95) [reply server/server.go:209]: To check validity of template_data

File: server/server_read_report_test.go [hunk 19]
```diff
@@ -118,47 +120,51 @@ func TestReadReportDBError(t *testing.T) {
...
+		}],
+		"fingerprints": [],
+		"skips": [],
+		"info": []
+}`
+	badHitRules := []types.ReportItem{
+		{
+			Module:       testdata.Rule1ID,
+			ErrorKey:     testdata.ErrorKey1,
+			TemplateData: json.RawMessage("not-json"),
```
[the remark this answers is not in the document]
Comment (bond95) [reply server/server_read_report_test.go:142]: Because it should be in parsed JSON format

[the remark this answers is not in the document]
Comment (Sergey1011010) [reply server/server_read_report_test.go:142]: Uhm.. Why can't you use just a string here?

[the remark this answers is not in the document]
Comment (Sergey1011010) [reply server/server_read_report_test.go:142]: we've got a function for that https://github.com/RedHatInsights/insights-operator-utils/blob/master/tests/helpers/strings.go#L71

File: server/server.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (Sergey1011010) [reply server/server.go:209]: we've got a function for that https://github.com/RedHatInsights/insights-operator-utils/blob/master/tests/helpers/strings.go#L71

[the remark this answers is not in the document]
Comment (bond95) [reply server/server.go:209]: oh, ok, thank you

File: server/server_read_report_test.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (bond95) [reply server/server_read_report_test.go:142]: Because type of `TemplateData` is  `json.RawMessage`

[the remark this answers is not in the document]
Comment (Sergey1011010) [reply server/server_read_report_test.go:142]: Hm, that's a weird choice for a type, but ok, probably there's a reason

File: consumer/processing.go [hunk 20]
```diff
@@ -241,5 +243,10 @@ func parseMessage(messageValue []byte) (incomingMessage, error) {
 		return deserialized, err
 	}
 
+	err = json.Unmarshal(*((*deserialized.Report)["reports"]), &deserialized.ParsedHits)
```
Comment (Sergey1011010) [inline consumer/processing.go:246]: It looks like C :D

File: openapi.json [hunk 21]
```diff
@@ -264,6 +264,103 @@
...
+              "format": "uuid"
+            }
+          },
+          {
+            "name": "userId",
+            "in": "path",
+            "required": true,
+            "schema": {
+              "type": "string",
+              "minLength": 36,
```
Comment (Sergey1011010) [inline openapi.json:300]: Is it correct? I remember that user id didn't have such constraints

File: server/server.go [hunk 22]
```diff
@@ -173,10 +167,67 @@ func (server *HTTPServer) readReportForCluster(writer http.ResponseWriter, reque
...
+
+	reportRule := types.RuleOnReport{
+		TemplateData: templateData,
+		Module:       ruleID,
+		ErrorKey:     errorKey,
+	}
+
+	reportRule = server.getFeedbackAndTogglesOnRule(clusterName, userID, reportRule)
+
+	if err != nil {
```
Comment (Sergey1011010) [inline server/server.go:225]: Something is wrong here, the same error is checked twice

File: server/server_test.go [hunk 23]
```diff
@@ -609,7 +609,7 @@ func TestRuleToggle(t *testing.T) {
 			assert.Equal(t, testdata.Rule1ID, toggledRule.RuleID)
 			assert.Equal(t, testdata.UserID, toggledRule.UserID)
 			assert.Equal(t, expectedState, toggledRule.Disabled)
-			if toggledRule.Disabled == storage.RuleToggleDisable {
+			if toggledRule.Disabled == true {
```
Comment (Sergey1011010) [inline server/server_test.go:612]: Maybe it's better to change type of the variable

File: storage/rule_toggle.go [hunk 24]
```diff
@@ -40,7 +40,7 @@ type ClusterRuleToggle struct {
 	ClusterID  types.ClusterName
 	RuleID     types.RuleID
 	UserID     types.UserID
-	Disabled   RuleToggle
+	Disabled   bool
```
Comment (Sergey1011010) [inline storage/rule_toggle.go:43]: Why did you change it to bool?

**Review by Sergey1011010 — CHANGES_REQUESTED** [review]

**Review by Sergey1011010 — APPROVED** [review]

**Review by tisnik — APPROVED** [review]
seems legit!
